src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -51,31 +65,94 @@ data ArithmeticCircuit a = ArithmeticCircuit
         -- ^ The order of variable assignments
         acRNG      :: StdGen
     } deriving (Generic, NFData)
+--}
+
+data Circuit a = Circuit
+    {
+        acSystem   :: Map Natural (Constraint a),
+        -- ^ The system of polynomial constraints
+        acInput    :: [Natural],
+        -- ^ The input variables
+        acWitness  :: Map Natural a -> Map Natural a,
+        -- ^ The witness generation function
+        acVarOrder :: Map (Natural, Natural) Natural,
+        -- ^ The order of variable assignments
+        acRNG      :: StdGen
+    } deriving (Generic, NFData)
+
+data ArithmeticCircuit n a = ArithmeticCircuit
+  { acCircuit :: Circuit a
+  , acOutput  :: Vector n Natural
+  } deriving (Generic, NFData)
+
+constraintSystem :: ArithmeticCircuit n a -> Map Natural (Constraint a)
+constraintSystem = acSystem . acCircuit
+
+inputVariables :: ArithmeticCircuit n a -> [Natural]
+inputVariables = acInput . acCircuit
+
+witnessGenerator :: ArithmeticCircuit n a -> Map Natural a -> Map Natural a
+witnessGenerator = acWitness . acCircuit
+
+varOrder :: ArithmeticCircuit n a -> Map (Natural, Natural) Natural
+varOrder = acVarOrder . acCircuit
 
 ----------------------------------- Circuit monoid ----------------------------------
 
-instance Eq a => Semigroup (ArithmeticCircuit a) where
-    r1 <> r2 = ArithmeticCircuit
+instance Eq a => Semigroup (Circuit a) where
+    c1 <> c2 =
+        Circuit
+           {
+               acSystem   = acSystem c1 `union` acSystem c2
+               -- NOTE: is it possible that we get a wrong argument order when doing `apply` because of this concatenation?
+               -- We need a way to ensure the correct order no matter how `(<>)` is used.
+           ,   acInput    = nub $ acInput c1 ++ acInput c2
+           ,   acWitness  = union <$> acWitness c1 <*> acWitness c2
+           ,   acVarOrder = acVarOrder c1 `union` acVarOrder c2
+           ,   acRNG      = mkStdGen $ fst (uniform (acRNG c1)) Haskell.* fst (uniform (acRNG c2))
+           }
+
+instance (Eq a, MultiplicativeMonoid a) => Monoid (Circuit a) where
+    mempty =
+        Circuit
+           {
+               acSystem   = empty,
+               acInput    = [],
+               acWitness  = insert 0 one,
+               acVarOrder = empty,
+               acRNG      = mkStdGen 0
+           }
+
+instance Eq a => Semigroup (ArithmeticCircuit n a) where
+    r1 <> r2 =
+        ArithmeticCircuit
+            {
+                acCircuit = acCircuit r1 <> acCircuit r2
+            ,   acOutput  = Z.zipWith max (acOutput r1) (acOutput r2)
```

Comment:
Previously, we prioritized the left circuit here. We should carefully check that it does not introduce new errors.

Reply:
If we replace the type of the state in `MonadBlueprint` instance with `Circuit`, we will not need `Monoid (ArithmeticCircuit n a)` instances at all, only the `Monoid (Circuit a)` ones

Reply:
Monoid instance and circuit in state/blueprint should have outputs n=0